src/coreComponents/linearAlgebra/multiscale/msrsb/MsrsbLevelBuilderCoupled.cpp

```diff
@@ -0,0 +1,338 @@
+      fineMatrix.multiplyPtAP( fine.m_selectors[blockIdx], fineBlock );
+    }
+
+    bool const updateBlock = m_builders[blockIdx]->updateProlongation( fineBlock );
+    CRSMatrixView< real64, globalIndex const > const block = m_prolongationBlocks[blockIdx].toViewConstSizes();
+
+    if( updateBlock )
+    {
+      GEOS_MARK_SCOPE( merge blocks );
+      m_builders[blockIdx]->prolongation().extract( block );
+      forAll< parallelHostPolicy >( block.numRows(),
+                                    [block = block.toViewConst(),
+                                     prolongation = m_localProlongation.toViewConstSizes(),
+                                     rowOffset]( localIndex const localRow )
+      {
+        auto const blockVals = block.getEntries( localRow );
+        auto const coupledVals = prolongation.getEntries( rowOffset + localRow );
+        GEOS_ASSERT_EQ( blockVals.size(), coupledVals.size() );
+        std::copy( blockVals.dataIfContiguous(),
+                   blockVals.dataIfContiguous() + blockVals.size(),
+                   coupledVals.dataIfContiguous() );
+      } );
+    }
+    update = update || updateBlock;
+    rowOffset += block.numRows();
+  }
+
+  if( update )
+  {
+    GEOS_MARK_SCOPE( create LAI object );
+    GEOS_LOG_RANK_0_IF( m_params.multiscale.debugLevel >= 2, GEOS_FMT( "[MsRSB] {}: updating coupled prolongation", m_name ) );
+    m_prolongation.create( m_localProlongation.toViewConst(),
+                           m_dofManager.numLocalDofs(),
+                           fineMatrix.comm() );
+  }
+
+  if( m_params.multiscale.debugLevel >= 4 )
+  {
+    m_prolongation.write( GEOS_FMT( "{}_P_{}.mtx", m_name, "conv" ), LAIOutputFormat::MATRIX_MARKET );
+  }
+
+  return update;
+}
+
+template< typename LAI >
+std::unique_ptr< PreconditionerBase< LAI > >
+MsrsbLevelBuilderCoupled< LAI >::makeCoarseSolver() const
+{
+  LinearSolverParameters params = m_params;
+  if( params.multiscale.coarseType == LinearSolverParameters::PreconditionerType::direct )
+  {
+    params.solverType = LinearSolverParameters::SolverType::direct;
+    return LAI::createSolver( params );
+  }
+  else if( params.multiscale.coarseType == LinearSolverParameters::PreconditionerType::block )
+  {
+    GEOS_ERROR_IF_NE_MSG( m_builders.size(), 2, "More than 2 blocks are not supported yet" );
+    auto solver = std::make_unique< BlockPreconditioner< LAI > >( params.block );
+    for( integer i = 0; i < static_cast< integer >( m_builders.size() ); ++i )
+    {
+      geos::DofManager::SubComponent comp{ params.block.subParams[i]->multiscale.fieldName, { m_builders[i]->numComp(), true } };
+      solver->setupBlock( m_params.block.order[i], { comp }, m_builders[i]->makeCoarseSolver() );
+      solver->setProlongation( m_params.block.order[i], m_selectors[i] );
+    }
+    return solver;
+  }
+  else
+  {
+    params.preconditionerType = params.multiscale.coarseType;
+    return LAI::createPreconditioner( params );
+  }
```

Comment:
just out of curiosity, how much work would it be to add hypre as an option here? So we could try a hybrid MsRSB+BoomerAMG @klevzoff @castelletto1

Reply:
I can't remember if I tested this thoroughly, but I believe this should already work - line 318 will create BoomerAMG (via `HyprePreconditioner`) as a coarse-level solver, with parameters specified in the XML (`amgXXX` attributes). One can also limit the number of multiscale levels to 1 or 2 (`maxLevels` attribute) to leave a sufficiently large coarse problem for AMG to work on.

Reply:
Nice, I'll take a look!